server/src/main/java/org/spine3/server/entity/AbstractEntity.java

```diff
@@ -0,0 +1,277 @@
+     *
+     * <p>Results of this method call are:
+     * <ul>
+     *   <li>The state object is set to the value produced by {@link #getDefaultState()}.
+     *   <li>The version number is set to zero.
+     *   <li>The {@link #whenModified} field is set to the system time of the call.
+     *   <li>The {@link #status} field is set to the default instance.
+     * </ul>
+     *
+     * <p>This method cannot be called from within {@code Entity} constructor because
+     * the call to {@link #getDefaultState()} relies on completed initialization
+     * of the instance.
+     */
+    @Override
+    protected void init() {
+        super.init();
+        setState(getDefaultState(), 0, getCurrentTime());
+        this.status = EntityStatus.getDefaultInstance();
+    }
+
+    /**
+     * Validates the passed state.
+     *
+     * <p>Does nothing by default. Aggregates may override this method to
+     * specify logic of validating initial or intermediate state.
+     *
+     * @param state a state object to replace the current state
+     * @throws IllegalStateException if the state is not valid
+     */
+    @SuppressWarnings({"NoopMethodInAbstractClass", "UnusedParameters"})
+    // Have this no-op method to prevent enforcing implementation in all sub-classes.
+    protected void validate(S state) throws IllegalStateException {
+        // Do nothing by default.
+    }
+
+    /**
+     * Validates and sets the state.
+     *
+     * @param state the state object to set
+     * @param version the entity version to set
+     * @param whenLastModified the time of the last modification to set
+     * @see #validate(S)
+     */
+    protected void setState(S state, int version, Timestamp whenLastModified) {
+        validate(state);
+        injectState(state);
+        setVersion(version, whenLastModified);
+    }
+
+    /**
+     * Sets status for the entity.
+     */
+    void setStatus(EntityStatus status) {
+        this.status = status;
+    }
+
+    /**
+     * Sets version information of the entity.
+     *
+     * @param version the version number of the entity
+     * @param whenLastModified the time of the last modification of the entity
+     */
+    protected void setVersion(int version, Timestamp whenLastModified) {
+        this.version = version;
+        this.whenModified = checkNotNull(whenLastModified);
+    }
+
+    /**
+     * Updates the state incrementing the version number and recording time of the modification.
+     *
+     * @param newState a new state to set
+     */
+    protected void incrementState(S newState) {
+        setState(newState, incrementVersion(), getCurrentTime());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getVersion() {
+        return version;
+    }
+
+    /**
+     * Advances the current version by one and records the time of the modification.
+     *
+     * @return new version number
+     */
+    protected int incrementVersion() {
+        ++version;
```

Comment:
It seems the `version` would look better as something like `AtomicInteger` to make the increments concurrency-friendly.
Maybe we could even create a `Version` class with synchronized methods. 

Just a thought.